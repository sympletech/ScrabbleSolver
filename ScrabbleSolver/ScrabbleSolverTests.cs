using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using ApprovalTests.Reporters;
using NUnit.Framework;

namespace ScrabbleSolver
{
    [TestFixture]
    [UseReporter(typeof(DiffReporter))]
    public class ScrabbleSolverTests
    {
        [Test]
        public void TwoTileTest()
        {
            string[] tiles =
                {
                    "a", "b"
                };
            var results = Solver.getCombinations(tiles.ToList());

            ApprovalTests.Approvals.VerifyAll(results, "foo"); 
        }
        [Test]
        public void TwoTileTest2()
        {
            string[] tiles =
                {
                    "c", "d"
                };
            var results = Solver.getCombinations(tiles.ToList());

            ApprovalTests.Approvals.VerifyAll(results, "foo");
        }
        [Test]
        public void TwoTileTest3()
        {
            string[] tiles =
                {
                    "a","b","c", "d"
                };
            var results = Solver.getCombinations(tiles.ToList());

            ApprovalTests.Approvals.VerifyAll(results, "foo");
        }
    }

    public static class Solver{
        public static IEnumerable<String> getCombinations(List<string> tiles)
        {
            var combinations = RecurseTiles(tiles);
            return combinations;
        }

        private static List<string> RecurseTiles(List<string> tiles)
        {
            List<string> combinations=new List<string>();
            for (int j = 0; j < tiles.Count(); j++)
            {
                combinations.Add(tiles[j]);
                var smallerList = tiles.Except(new List<string>(){tiles[j]});
                // smallerList.Remove(smallerList[j]);
                if (smallerList.Count() > 0)
                {
                    var therest = RecurseTiles(smallerList.ToList());
                    foreach (var combo in therest)
                    {
                        combinations.Add(tiles[j] + combo);
                    }
                }
            }
            return combinations;

        }
    }
}
