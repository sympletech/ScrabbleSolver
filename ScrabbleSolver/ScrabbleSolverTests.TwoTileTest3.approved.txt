foo[0] = a
foo[1] = ab
foo[2] = abc
foo[3] = abcd
foo[4] = abd
foo[5] = abdc
foo[6] = ac
foo[7] = acb
foo[8] = acbd
foo[9] = acd
foo[10] = acdb
foo[11] = ad
foo[12] = adb
foo[13] = adbc
foo[14] = adc
foo[15] = adcb
foo[16] = b
foo[17] = ba
foo[18] = bac
foo[19] = bacd
foo[20] = bad
foo[21] = badc
foo[22] = bc
foo[23] = bca
foo[24] = bcad
foo[25] = bcd
foo[26] = bcda
foo[27] = bd
foo[28] = bda
foo[29] = bdac
foo[30] = bdc
foo[31] = bdca
foo[32] = c
foo[33] = ca
foo[34] = cab
foo[35] = cabd
foo[36] = cad
foo[37] = cadb
foo[38] = cb
foo[39] = cba
foo[40] = cbad
foo[41] = cbd
foo[42] = cbda
foo[43] = cd
foo[44] = cda
foo[45] = cdab
foo[46] = cdb
foo[47] = cdba
foo[48] = d
foo[49] = da
foo[50] = dab
foo[51] = dabc
foo[52] = dac
foo[53] = dacb
foo[54] = db
foo[55] = dba
foo[56] = dbac
foo[57] = dbc
foo[58] = dbca
foo[59] = dc
foo[60] = dca
foo[61] = dcab
foo[62] = dcb
foo[63] = dcba
